### PR TITLE
Company Ai: Purchase Vehicle

### PR DESCRIPTION
**Pull request by duncanspumpkin**
There a good few issues in the vanilla version of this. I'm somewhat surprised it matches the replays.

### CODE REVIEW

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 1]
```diff
@@ -292,10 +452,10 @@ namespace OpenLoco
     // 0x00430971
     static void aiThinkState1(Company& company)
     {
-        company.var_2578++;
-        if (company.var_2578 < 60)
+        company.activeThoughtId++;
+        if (company.activeThoughtId < 60)
```
Comment (ZehMatt) [inline src/OpenLoco/src/World/CompanyAi.cpp:456]: Should also have a constant for max thoughts.

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 2]
```diff
@@ -74,15 +81,168 @@ namespace OpenLoco
...
+        for (auto i = 0; i < thought.var_45; ++i)
+        {
+            GameCommands::VehicleCreateArgs createArgs{};
+            createArgs.vehicleId = trainHeadId;
+            createArgs.vehicleType = thought.var_46[i];
+            auto res = GameCommands::doCommand(createArgs, GameCommands::Flags::apply);
+            if (res == GameCommands::FAILURE)
+            {
+                // If we can't create a vehicle try create a free vehicle!
+                res = GameCommands::doCommand(createArgs, GameCommands::Flags::apply | GameCommands::Flags::noPayment);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/CompanyAi.cpp:114]: Wait what?! 😄

File: src/OpenLoco/src/World/CompanyAi.h [hunk 3]
```diff
@@ -72,18 +72,18 @@ namespace OpenLoco
...
-        uint8_t var_45;              // 0x4ED size of var_46
-        uint8_t pad_46[0x66 - 0x46]; // array of uint16_t object id's unsure of size
-        EntityId var_66[8];          // 0x50E
-        currency32_t var_76;         // 0x51E
+        uint8_t cargoType;    // 0x42 0x4EA
+        uint8_t var_43;       // 0x4EB
+        uint8_t numVehicles;  // 0x44 0x4EC size of var_66
+        uint8_t var_45;       // 0x4ED size of var_46
+        uint16_t var_46[16];  // 0x4EF array of uint16_t object id
+        EntityId vehicles[8]; // 0x66 0x50E see also numVehicles for current size
```
Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/CompanyAi.h:85]: We should probably reset these in the usurp company assets cheat 👀

Comment (AaronVanGeffen) [inline src/OpenLoco/src/World/CompanyAi.h:85]: What's the meaning behind the double offsets here?

Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.h:85]: I find that sometimes CS access this structure from a pointer to the base of it and sometimes from the base of the company. So the first number is the structure the second the company.

Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.h:85]: Yes I think that might be the source of some of the issues with the cheat.

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 2 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:114]: cheaters!

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 4]
```diff
@@ -74,15 +81,168 @@ namespace OpenLoco
...
+            return PurchaseVehicleResult::failure;
+        }
+
+        if (thought.numVehicles >= thought.var_43)
+        {
+            return PurchaseVehicleResult::allVehiclesPurchased;
+        }
+
+        EntityId trainHeadId = EntityId::null;
+        for (auto i = 0; i < thought.var_45; ++i)
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/CompanyAi.cpp:105]: can `var_45` be named?

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 5]
```diff
@@ -74,15 +81,168 @@ namespace OpenLoco
...
+        {
+            GameCommands::VehicleCreateArgs createArgs{};
+            createArgs.vehicleId = trainHeadId;
+            createArgs.vehicleType = thought.var_46[i];
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/CompanyAi.cpp:109]: looks like `var_46` is the list of vehicles the ai is thinking about?

File: src/OpenLoco/src/World/CompanyAi.cpp [hunk 6]
```diff
@@ -74,15 +81,168 @@ namespace OpenLoco
...
+        {
+            return PurchaseVehicleResult::success;
+        }
+        if (thought.var_43 > 1)
+        {
+            // Why??
+            GameCommands::VehicleOrderSkipArgs skipArgs{};
+            skipArgs.head = trainHeadId;
+            GameCommands::doCommand(skipArgs, GameCommands::Flags::apply);
+            GameCommands::doCommand(skipArgs, GameCommands::Flags::apply);
```
Comment (LeftofZen) [inline src/OpenLoco/src/World/CompanyAi.cpp:243]: what does this block actually do?

**Review by LeftofZen — APPROVED** [review]
playtested for about 5 in-game years, AI seemed perfectly capable of building vehicles as expected

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:105]: I like to leave naming of these until i'm really sure there aren't any other uses. I'm not 100% sure yet.

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 5 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:109]: I like to leave naming of these until i'm really sure there aren't any other uses. I'm not 100% sure yet.

File: src/OpenLoco/src/World/CompanyAi.cpp (hunk 6 above)
Comment (duncanspumpkin) [reply src/OpenLoco/src/World/CompanyAi.cpp:243]: I've no idea :(

**Pull request merged**